lib/components/primitive-components/Trace/get-obstacles-for-trace.ts

```diff
@@ -23,7 +23,49 @@ export const getSchematicObstaclesForTrace = (trace: Trace): Obstacle[] => {
   const obstacles: Obstacle[] = []
 
   // Add obstacles from components and ports
+  let minX = Infinity,
+    minY = Infinity,
+    maxX = -Infinity,
+    maxY = -Infinity
   for (const elm of db.toArray()) {
+    // Track bounds for border obstacle
+    let cx: number | undefined,
+      cy: number | undefined,
+      w: number | undefined,
+      h: number | undefined
+    if (elm.type === "schematic_component") {
+      cx = elm.center?.x
+      cy = elm.center?.y
+      w = elm.size?.width
+      h = elm.size?.height
+    } else if (elm.type === "schematic_box") {
+      cx = elm.x
+      cy = elm.y
+      w = elm.width
+      h = elm.height
+    } else if (elm.type === "schematic_port") {
+      cx = elm.center?.x
+      cy = elm.center?.y
+      w = 0.2
+      h = 0.2
+    } else if (elm.type === "schematic_text") {
+      cx = elm.position?.x
+      cy = elm.position?.y
+      w = (elm.text?.length ?? 0) * 0.1
+      h = 0.2
+    }
+    if (
+      typeof cx === "number" &&
+      typeof cy === "number" &&
+      typeof w === "number" &&
+      typeof h === "number"
+    ) {
+      minX = Math.min(minX, cx - w / 2)
+      maxX = Math.max(maxX, cx + w / 2)
+      minY = Math.min(minY, cy - h / 2)
+      maxY = Math.max(maxY, cy + h / 2)
+    }
```

Comment:
You should isolate this into a separate function eg getBoundsForSchematicElement, where possible and ergonomic we should avoid extending these already large class files